StandaloneMmPkg/Core/StandaloneMmCore.c

```diff
@@ -440,33 +500,77 @@ MmEntryPoint (
   // Check to see if this is a Synchronous MMI sent through the MM Communication
   // Protocol or an Asynchronous MMI
   //
-  if (gMmCorePrivate->CommunicationBuffer != 0) {
-    //
-    // Synchronous MMI for MM Core or request from Communicate protocol
-    //
-    if (!MmIsBufferOutsideMmValid ((UINTN)gMmCorePrivate->CommunicationBuffer, gMmCorePrivate->BufferSize)) {
-      //
-      // If CommunicationBuffer is not in valid address scope, return EFI_INVALID_PARAMETER
-      //
-      gMmCorePrivate->CommunicationBuffer = 0;
-      gMmCorePrivate->ReturnStatus        = EFI_INVALID_PARAMETER;
-    } else {
-      CommunicateHeader           = (EFI_MM_COMMUNICATE_HEADER *)(UINTN)gMmCorePrivate->CommunicationBuffer;
-      gMmCorePrivate->BufferSize -= OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data);
-      Status                      = MmiManage (
-                                      &CommunicateHeader->HeaderGuid,
-                                      NULL,
-                                      CommunicateHeader->Data,
-                                      (UINTN *)&gMmCorePrivate->BufferSize
-                                      );
+  if ((mMmCommunicationBuffer != NULL) && (mInternalCommBufferCopy != NULL)) {
+    CommunicationInOut = (COMMUNICATION_IN_OUT *)(UINTN)mMmCommunicationBuffer->CommunicationInOut;
+    if (CommunicationInOut->IsCommBufferValid) {
       //
-      // Update CommunicationBuffer, BufferSize and ReturnStatus
-      // Communicate service finished, reset the pointer to CommBuffer to NULL
+      // Synchronous MMI for MM Core or request from Communicate protocol
       //
-      gMmCorePrivate->BufferSize         += OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data);
-      gMmCorePrivate->CommunicationBuffer = 0;
-      gMmCorePrivate->ReturnStatus        = (Status == EFI_SUCCESS) ? EFI_SUCCESS : EFI_NOT_FOUND;
+      if (!MmIsBufferOutsideMmValid (
+             (UINTN)mMmCommunicationBuffer->FixedCommBuffer,
+             mMmCommunicationBuffer->FixedCommBufferSize
+             ))

```

Comment:
You already validated the FixedCommBuffer location in MmCoreEntryPoint. It's not needed to validate again.